compiler/src/dotty/tools/dotc/tastyreflect/QuoteContextState.scala

```diff
@@ -0,0 +1,22 @@
+package dotty.tools.dotc.tastyreflect
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.util.Property
+
+object QuoteContextState {
+
+  val NextIndex = new Property.Key[scala.runtime.IntRef]
+
+  def init(ctx: Context): Context =
+    ctx.fresh.setProperty(NextIndex, new scala.runtime.IntRef(0))
+
+  def nextIndex()(given ctx: Context): Int = {
+    val ref = ctx.property(NextIndex).get
+    val ret = ref.elem
+    ref.elem = ret + 1
+    ret
+  }
+}
```

Comment:
Maybe reuse [FreshNameCreator](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/FreshNameCreator.scala) from the compiler?

Reply:
It would have a global indexing which would make any debugging harder as it would keep undesired state from one macro expansion to the other.

Reply:
If a macro contains many usage of `let`, then it faces the same problem. 

`FreshNameCreator` is better in the sense that it keeps different counters for different names, instead of using one global counter.

Reply:
But it uses a global counter across all files in the compilation run and each macro expansion site which is what we are trying to avoid.

Reply:
It's possible to call `def setFreshNames(freshNames: FreshNameCreator)` on `Context` before macro expansion.

Reply:
Maybe a could try to create a fresh `FreshNameCreator` for each macro expansion site.